Composer/packages/webchat-client/src/WebChatPanel.tsx

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import React, { useMemo, useEffect, useState, useRef } from 'react';
+import ReactWebChat, { createStyleSet } from 'botframework-webchat';
+import formatMessage from 'format-message';
+import { createStore as createWebChatStore } from 'botframework-webchat-core';
+import { CommunicationColors, NeutralColors } from '@uifabric/fluent-theme';
+
+import webChatStyleOptions from './utils/webChatTheme';
+import ConversationService, { ChatData, BotSecrets } from './utils/ConversationService';
+import { WebChatHeader } from './WebChatHeader';
+
+const BASEPATH = process.env.PUBLIC_URL || 'http://localhost:3000/';
+
+export interface WebChatPanelProps {
+  botUrl: string;
+  secrets: BotSecrets;
+  /** Directline host url. By default, set to Composer host url. */
+  directlineHostUrl?: string;
+  botName: string;
+  projectId: string;
+  isWebChatPanelVisible: boolean;
+  activeLocale: string;
+  appLifecycleHandler: any;
+  openBotInEmulator: (projectId: string) => void;
+}
+
+export const WebChatPanel: React.FC<WebChatPanelProps> = ({
+  projectId,
+  botUrl,
+  secrets,
+  directlineHostUrl = BASEPATH,
+  botName,
+  isWebChatPanelVisible,
+  openBotInEmulator,
+  activeLocale,
+  appLifecycleHandler,
+}) => {
+  const [chats, setChatData] = useState<Record<string, ChatData>>({});
+  const [currentConversation, setCurrentConversation] = useState<string>('');
+  const conversationServiceRef = useRef<ConversationService>(new ConversationService(directlineHostUrl));
```

Comment:
`useMemo` is a better hook for this pattern.

```suggestion
  const conversationService = useMemo(() => new ConversationService(directlineHostUrl), [directlineHostUrl]);
```